### PR TITLE
Consistently format urls in history.

### LINKED ISSUE #2903: URL percent-encoding in the SQL history
There are couple of problems:

- The [add_url](https://github.com/qutebrowser/qutebrowser/blob/49b858e3599a361cf5c994358c6b189dddfb522c/qutebrowser/browser/history.py#L162) method does percent-encoding, but the [parse_entry](https://github.com/qutebrowser/qutebrowser/blob/49b858e3599a361cf5c994358c6b189dddfb522c/qutebrowser/browser/history.py#L192) method does not. So the URLs which were imported from the text file are not percent-encoded, but the newly added URLs are, which leads to "duplicates" if I open an old URL.
- The old history completion called [entry.url.toDisplayString()](https://github.com/qutebrowser/qutebrowser/pull/2295/files#diff-d0e110fc69342b1ed59bb5c1a96b0234L101) when the URL was added to the completion. Now the completion shows URLs as they are in the database, which is inconsistent (see the previous point) and a visual regression (I want pretty URLs in the completion).

cc @rcorre

### PR DESCRIPTION
**Pull request by rcorre**
Encode urls that are inserted into the history, but do not encode urls
for completion (other than removing passwords).
Also ensure that urls read from the history text file are formatted
consistenly with those added while browsing.

Fixes #2903.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/2905)

### CODE REVIEW

**Comment by The-Compiler** [comment]
@lahwaacz can you test and/or review this?

**Comment by lahwaacz** [comment]
Haven't tested yet, but I have couple of questions:

- Why does the URL in the History table have to be encoded? With the previous textual backend I'd say it was to allow simpler parsing, but in SQL spaces or other special characters don't matter. You could have everything unencoded for coherence/simplicity.
- The changes still don't handle the "duplicates", do they? The simplest migration I can think of is to drop the CompletionHistory table and rebuild it from the History table again. I think this should be possible anyway if users were to be allowed to modify History as they like (e.g. remove old rows) and expect the completion to be consistent.

**Comment by rcorre** [comment]
> You probably wanted to comment on the pull request rather than here :confused: (though we could still copy-paste our comments and delete the ones here...)

Yup, I did.

> As for the "duplicates", I already have a spa%20ce-type URL in the CompletionHistory table, so adding spa ce does not hit any primary key constraint and both URLs will coexist forever. The only difference the changes make is that the spa ce URL will appear above spa%20ce, because it has the more recent timestamp.

I see what you're saying. This will work for those who go from the old history directly to this commit or later. It won't help those who are already running the latest prerelease qutebrowser, because their history is already screwed up. I think the best solution there is just to run `:debug-dump-history ~/.local/share/qutebrowser/history`, `mv ~/.local/share/qutebrowser/history.sqlite{,.bak}`, and then restart qutebrowser. This should start you with fresh history and completion tables where nothing is encoded.

NOTE: I haven't actually tried the above. Proceed with caution.

> I think this should be possible anyway if users were to be allowed to modify History as they like (e.g. remove old rows) and expect the completion to be consistent.

So something like a `regenerate-history` command to be run after manually editing `history.sqlite`?


You mean

**Comment by lahwaacz** [comment]
> I think the best solution there is just to run `:debug-dump-history ~/.local/share/qutebrowser/history`, mv `~/.local/share/qutebrowser/history.sqlite{,.bak}`, and then restart qutebrowser.

IMHO there should be migrations even on git master (otherwise you can expect similar bug reports any time in the future), but this strategy should indeed work. Though I've just tried the `:debug-dump-history` command and it does not fully-encode the URLs, so the subsequent import would fail to parse it correctly.

> So something like a `regenerate-history` command to be run after manually editing `history.sqlite`?

No command please, make it automatic. If users modify the `History` table, they should also drop `CompletionHistory` and then qutebrowser should automatically regenerate `CompletionHistory` from the `History` table.

By the way, this also allows us to do the migration without the intermediate text file: we would just drop `CompletionHistory` and qutebrowser would rebuild it &ndash; of course __explicitly decoding the URLs__, because the `History` table does not use a single URL format anymore.

**Comment by The-Compiler** [comment]
I think it'd be better to use `FullyEncoded` for the table FWIW. `QUrl::toDisplayString` uses `PrettyDecoded` by default, which is documented as "may also change from Qt release to Qt release".

It's probably okay to add the display string to `CompletionHistory` for performance reasons, but in `History` it'd be better to have URLs in a format which definitely isn't going to change between different Qt releases.

**Comment by rcorre** [comment]
Now I've got encoded urls for `History`, unencoded for `HistoryCompletion`, and auto-regeneration of `HistoryCompletion` if it is dropped.

File: qutebrowser/browser/history.py [hunk 1]
```diff
@@ -71,6 +73,17 @@ def __repr__(self):
     def __contains__(self, url):
         return self._contains_query.run(val=url).value()
 
+    def _rebuild_completion(self):
+        data = {'url': [], 'title': [], 'last_atime': []}
+        # select the latest entry for each url
+        q = sql.Query('SELECT url, title, max(atime) AS atime FROM History '
+                      'WHERE NOT redirect GROUP BY url')
+        for entry in q.run():
+            data['url'].append(entry.url)
```
Comment (lahwaacz) [inline qutebrowser/browser/history.py]: Shouldn't this use `QUrl::toDisplayString`?

**Comment by rcorre** [comment]
oops, fixed!

**Comment by The-Compiler** [comment]
We should probably still drop the HistoryCompletion database once now, to fix up old entries?

To track whether that's done (and future schema migrations), we could use the [user_version pragma](https://sqlite.org/pragma.html#pragma_user_version):

```
>>> from qutebrowser.misc import sql
>>> sql.init(':memory:')
>>> sql.Query('pragma user_version;').run().value()
0
>>> sql.Query('pragma user_version = 1;').run()
<qutebrowser.misc.sql.Query object at 0x7fc7fff82198>
>>> sql.Query('pragma user_version;').run().value()
1
```

So if it's 0, we drop the `HistoryCompletion` table once and let it re-generate, and either way we set it to `1` after init/migrations are completed.

Not sure how that'd fit best into the code though, as the pragma is global, and not per-table.

**Comment by rcorre** [comment]
I think it is fine to use the global pragma. If not, we could maintain a table of table-versions, but that seems like overkill to me, and not a big issue to change later if needed.

**Comment by The-Compiler** [comment]
Looks good from a quick look - @lahwaacz what do you think?

**Comment by lahwaacz** [comment]
I just got an uncaught exception when I run the new code with my current `history.sql` database:

```
Traceback (most recent call last):
  File "./qutebrowser.py", line 29, in <module>
    sys.exit(qutebrowser.qutebrowser.main())
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/qutebrowser.py", line 177, in main
    return app.run(args)
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/app.py", line 113, in run
    init(args, crash_handler)
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/app.py", line 139, in init
    _init_modules(args, crash_handler)
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/app.py", line 437, in _init_modules
    history.init(qApp)
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/browser/history.py", line 320, in init
    history = WebHistory(parent=parent)
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/browser/history.py", line 59, in __init__
    self._rebuild_completion()
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/browser/history.py", line 92, in _rebuild_completion
    self.completion.insert_batch(data)
  File "/home/lahwaacz/Bbox/pg/python3/qutebrowser/qutebrowser/misc/sql.py", line 232, in insert_batch
    q.lastQuery(), q.lastError().text()))
qutebrowser.misc.sql.SqlException: Failed to exec query "INSERT INTO CompletionHistory (url, title, last_atime) values(:url, :title, :last_atime)": "UNIQUE constraint failed: CompletionHistory.url Unable to fetch row"
```

I think it's probably due to [this line](https://github.com/qutebrowser/qutebrowser/pull/2905/commits/8c6133e29d3e6ac3489f2d5d4456e5527830613d#diff-01b81c150609cada156846c5ae680b9bR82) yielding the same key for URLs such as `spa%20ce` and `spa ce`, so inserting the next value fails. You can probably fix it with some form of an [UPSERT](https://stackoverflow.com/questions/15277373/sqlite-upsert-update-or-insert) query. Or, since we know that the table is initially empty in the rebuild, you could deduplicate the batch directly in Python...

**Comment by rcorre** [comment]
Should be fixed now. We also have to order the query now, as if two entries end up with the same url we need the larger atime to be inserted last.

**Comment by lahwaacz** [comment]
Seems to be working as expected now :+1:

**Comment by rcorre** [comment]
@The-Compiler you good with merging this?

**Comment by The-Compiler** [comment]
Seems to work fine, so let's merge it - thanks!

**Pull request merged**